pkg/proxy/iptables/number_generated_rules_test.go

```diff
@@ -0,0 +1,391 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package iptables
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	discovery "k8s.io/api/discovery/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	iptablestest "k8s.io/kubernetes/pkg/util/iptables/testing"
+	netutils "k8s.io/utils/net"
+	utilpointer "k8s.io/utils/pointer"
+)
+
+// kube-proxy generates iptables rules to forward traffic from Services to Endpoints
+// kube-proxy uses iptables-restore to configure the rules atomically, however,
+// this has the downside that large number of rules take a long time to be processed,
+// causing disruption.
+// There are different parameters than influence the number of rules generated:
+// - ServiceType
+// - Number of Services
+// - Number of Endpoints per Service
+// This test will fail when the number of rules change, so the person
+// that is modifying the code can have feedback about the performance impact
+// on their changes. It also runs multiple number of rules test cases to check
+// if the number of rules grows linearly.
+func TestNumberIptablesRules(t *testing.T) {
+	testCases := []struct {
+		name                string
+		epsFunc             func(eps *discovery.EndpointSlice)
+		svcFunc             func(svc *v1.Service)
+		services            int
+		epPerService        int
+		expectedFilterRules int
+		expectedNatRules    int
+	}{
+		{
+			name:                "0 Services 0 EndpointsPerService - ClusterIP",
+			services:            0,
+			epPerService:        0,
+			expectedFilterRules: 3,
+			expectedNatRules:    5,
+		},
+		{
+			name:                "1 Services 0 EndpointPerService - ClusterIP",
+			services:            1,
+			epPerService:        0,
+			expectedFilterRules: 4,
+			expectedNatRules:    5,
+		},
+		{
+			name:                "1 Services 1 EndpointPerService - ClusterIP",
+			services:            1,
+			epPerService:        1,
```

Comment:
because the real difference is between 1 and 2, I might add a case for 2 endpoints, too. :)